### PR TITLE
test: add test environment to run tests against local solana

### PR DESCRIPTION
**Pull request by olgkv**
- Add test environment to run tests against `local solana network`, using solana cli tool: `solana-test-validator`
- Add External Price Account test, which is run against `local solana network`
- Add Github Actions, which are allow to run tests after pull request

### CODE REVIEW

File: test/utils.ts
Comment (kurpav) [inline test/utils.ts:7]: please rename it to `LOCAL_NETWORK`

**Comment by kurpav** [comment]
@olgkv Good job 👍 
Please add a description to this PR

File: .github/workflows/pull_request.yml
Comment (b2kdaman) [inline .github/workflows/pull_request.yml:25]: I global install necessary?

Comment (b2kdaman) [reply .github/workflows/pull_request.yml:25]: Otherwise lgtm

Comment (olgkv) [reply .github/workflows/pull_request.yml:25]: I think it's ok, because it runs only in docker image on github

**Review by thlorenz — CHANGES_REQUESTED** [review]
Requested a few changes via a call.

**Comment by thlorenz** [comment]
@olgkv have a look at [amman](https://www.npmjs.com/package/@metaplex-foundation/amman), a package I just created to aid in setting up the validator and which will include convenience actions to be used for testing in the future.

All you have to do is place a `.ammnrc.js` config in the root of the project and then run `amman validator`.

The config would be a combination of the below two (I spit it into base and inheritor since the mpl repo has multiple packages).

### Base `.ammanrc.js`

```js
// @ts-check
'use strict';
const path = require('path');

const localDeployDir = path.join(__dirname, 'target', 'deploy');

const programIds = {
  metadata: 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s',
  vault: 'vau1zxA2LbssAUEF7Gpw91zMM1LvXrvpzJtmZ58rPsn',
  auction: 'auctxRXPeJoc4817jDhf4HbjnhEcr1cCXenosMhK5R8',
  metaplex: 'p1exdMJcjVao65QdewkaZRUnU6VPSXhus9n2GzWfh98',
};

function localDeployPath(programName) {
  return path.join(localDeployDir, `${programName}.so`);
}
const programs = {
  metadata: { programId: programIds.metadata, deployPath: localDeployPath('mpl_token_metadata') },
  vault: { programId: programIds.vault, deployPath: localDeployPath('mpl_token_vault') },
  auction: { programId: programIds.auction, deployPath: localDeployPath('mpl_auction') },
  metaplex: { programId: programIds.mpl, deployPath: localDeployPath('mpl_metaplex') },
};

const validator = {
  verifyFees: true,
};

module.exports = {
  programs,
  validator,
};
```

### TokenMetadata `.ammanrc.js`

```js
'use strict';
// @ts-check
const base = require('../../.ammanrc.js');

const validator = { ...base.validator, programs: [base.programs.metadata] };
module.exports = { validator };
```

LMK if you need help setting this up.

File: .github/workflows/pull_request.yml
Comment (thlorenz) [reply .github/workflows/pull_request.yml]: I'd also opt for removing this and only add this if it breaks (we're other github actions that also depend on yarn and don't need this install).

File: package.json
Comment (thlorenz) [inline package.json:38]: Don't use `npx` in production or anywhere please as it is basically the same as doing `curl <some-url> | bash`
You already added `amman` as a `devDependency` (which also allows it to be cached for CI and avoids devs re-installing it over and over even if yarn ends up using local cache). 
Therefore you can remove `npx`.

Also you're depending on devs to have `npx` installed which is another problem ...

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:18]: It is usually advised to not change dir during a script or app run.
Instead just resolve all paths provided to tools and give them full path so it doesn't matter from which directory you run.

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:23]: Please use a library from npm like `request` or `fetch` to perform this.
Never depend on any external tooling like `wget` to be installed on the dev's machine or CI environment.

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:27]: ditto regarding the `chdir`. I'm pretty sure there must be a way to provide the project directory to cargo. If that option does not exist then I suppose `chdir` is fine here, but then you should store the current directory and change back to it after you're done. Otherwise things get confusing very quickly.

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:24]: ditto regarding external tools .. please use Node.js zip library instead, multiple options exist as outlined [here](https://stackoverflow.com/a/46967462/97443)

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:32]: ditto chdir.

File: test/utils.ts
Comment (thlorenz) [inline test/utils.ts:110]: Given this runs against local validator now why do we still need to match the feepayer in devnet?
You could just generate a random keypair each time ... it shouldn't matter as your code + tests should work no matter what that key is.

File: test/utils.ts
Comment (thlorenz) [inline test/utils.ts:7]: This is fine, but to avoid duplication and make sure you're connecting to the correct URL you should just do `import { LOCALHOST as LOCAL_NETWORK } from '@metaplex-foundation/amman'`.

File: .gitignore
Comment (thlorenz) [inline .gitignore:23]: Do we need this? Amman writes the ledger to a `/tmp` dir so this folder shouldn't be created.

File: .gitignore
Comment (thlorenz) [inline .gitignore:24]: This is also no longer needed as I pointed out previously ... config is written to tmp dir and cleaned up by amman.

**Review by thlorenz — CHANGES_REQUESTED** [review]
Found some issues especially regarding the build setup depending on external tools.
Please fix and LMK when done so I can review again.

Also please DO address the points I'm raising before asking for another review, I feel like I'm commenting on some things (like the `solana-validator.yml` in the .gitignore) a second time.

File: test/utils.ts
Comment (olgkv) [reply test/utils.ts:110]: I prefer to leave it how it is for now, other way I have to refactor many tests, which I'm going to do eventually step by step

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:10]: `s/repostitoryDir/repositoryDir`

File: test/utils.ts
Comment (thlorenz) [inline test/utils.ts:134]: Just pleas use JSON parse here instead of implementing this manually after you read the file.
`require` does this all in one step and will even keep the parsed JSON in the `require.cache`.

So this whole function would be be something along the lines of:

```js
const arr = require(keypairPath);
const u8Array = Uint8Array.from(arr);
return Keypair.fromSecretKey(u8Array);
```

**Review by thlorenz — CHANGES_REQUESTED** [review]
Getting closer, please fix the final points/nits I raised.

File: test/setup/build.ts
Comment (olgkv) [reply test/setup/build.ts]: I guess you're using vim:)

**Comment by thlorenz** [comment]
You resolved the issues I had, but I'm confused WRT to the github workflow .. it doesn't seem to complete .. ever?
> Pull Request _Expected — Waiting for status to be reported_

I'd, as the changes are fine, but I don't want yo to merge this until we figure out why those checks aren't running properly.

Nothing should get merged with passing checks and this PR adds a github workflow that runs integration tests as part of a workflow. So we need to see those passing before merging.

Does anyone know why they're stuck?

**Review by b2kdaman — DISMISSED** [review]

**Review by zaxozhu — DISMISSED** [review]

File: .github/workflows/test-integration.yml
Comment (kurpav) [inline .github/workflows/test-integration.yml:7]: I guess we should run tests for each PR

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:24]: if it is typescript please use types in this file

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:28]: please use async/await with try/catch

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:43]: I thought you removed `chdir`

**Review by kurpav — CHANGES_REQUESTED** [review]

File: test/setup/build.ts
Comment (thlorenz) [reply test/setup/build.ts:43]: I told Oleg to remove all avoidable `chdir` calls .. I'm assuming this is unavoidable? Otherwise I'd opt for removing that as well.

File: .github/workflows/test-integration.yml
Comment (thlorenz) [reply .github/workflows/test-integration.yml:7]: On each PR to master  .. but that may be why those aren't running since in this repo it's `main`. So we should adapt this.
I hope this isn't the reason why the other workflow isn't triggered either?

File: test/setup/build.ts
Comment (olgkv) [reply test/setup/build.ts:43]: I don't know how to provide cargo where it should look out for configuration

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:33]: Please only log errors on `stderr` via `console.error`.
Did you actually check what happens here on errors? If I understand correctly you log the error but then continue with the script which is not correct, i.e. if  a step fails you should exit with an error code `!=0`.

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:38]: Ditto reqarding error logging, but you'd never even get here since you already log and then ignore the error inside the `download` method.

Please chose _one_ place to handle the error and then don't just log it but think about what the consequence should be and as I explained above you should probably exit with an error code.

File: test/utils.ts
Comment (thlorenz) [inline test/utils.ts:1]: Please just remove unused imports instead of leaving stray comments

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:33]: please use `try`/`catch` instead of `then`/`catch`

**Review by thlorenz — CHANGES_REQUESTED** [review]
Nice, I see the integration tests now actually are running 😄 .
There are still some issues that I pointed out.

Also could someone please figure out why the _Pull Request_ workflow isn't running? We don't know for instance if the lint check passes and so on. It most likely also would have caught the missing TypeScript types etc.

**Review by kurpav — CHANGES_REQUESTED** [review]

File: .github/workflows/test-integration.yml
Comment (thlorenz) [inline .github/workflows/test-integration.yml:28]: Just realizing you're linting here which shouldn't be part of the _Integration Test_ as it is covered by the [pull request](https://github.com/metaplex-foundation/js/actions/runs/1591894692/workflow) workflow (assuming it runs).

File: .github/workflows/pull_request.yml
Comment (thlorenz) [inline .github/workflows/pull_request.yml:1]: Hold on why was this deleted?
We should still have something that builds + runs unit tests (without solana) and lints.
I may have told you to rename it to _what it does_ , i.e. "Build Lint and Unit Test" instead of when it runs (pull request), but you shouldn't just delete it.

This may be why it doesn't run

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:32]:
```suggestion
      const response = await axios({ method: 'get', url: path, responseType: 'stream' });
```

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:34]:
```suggestion
        await done(writer);
```

File: test/setup/build.ts
Comment (kurpav) [inline test/setup/build.ts:35]:
```suggestion
```

File: test/setup/build.ts
Comment (thlorenz) [inline test/setup/build.ts:48]: Very confusing that this is upper case as it looks like an error constructor (but it is just a function).
Please change naming to make this easier to understand, i.e. `printErrorAndExit`.

File: test/metadata.test.ts
Comment (thlorenz) [inline test/metadata.test.ts:5]: I don't understand the reasoning here.
How should a test become outdated if you change nothing about the SDK code itself?
At any rate just skipping without an explanation isn't a good solution. Ideally the test would be adapted and thus at the very least you should add a `TODO` here that in a subsequent PR this test will be fixed and re-enabled.

If it is entirely outdated then it should be removed eventually, but also then you'd need to figure out why it is outdated and if you can adapt it instead of removing it entirely.

**Review by thlorenz — CHANGES_REQUESTED** [review]

File: test/metadata.test.ts
Comment (olgkv) [reply test/metadata.test.ts:5]: The test failed every time, and  it runs in 'devnet'. There was an idea to rewrite it against local solana, but in this way the test will be the identical as  [createMetadataAndME test]( https://github.com/metaplex-foundation/js/blob/main/test/actions/createMetadataAndME.test.ts), so we have to create metadata and master edition in the both tests to verify them. So I think it's quite unnecessary actions  to duplicate tests, and I'm going to remove on the next iteration

Comment (thlorenz) [reply test/metadata.test.ts:5]: If the test is an actual duplicate then it should be removed eventually (not skipped).
This could be done in a separate PR that also points to the duplicate test explaining how they cover the same use case.

Comment (olgkv) [reply test/metadata.test.ts:5]: I absolutely agree, this will be removed in a separate PR

**Review by thlorenz — APPROVED** [review]
Approving with the understanding that the skipped test will be addressed.

**Review by kurpav — APPROVED** [review]

**Pull request merged**
